knowledge/data-sources/notion/src/index.ts

```diff
@@ -1,222 +1,282 @@
+    }
+    return title;
 }
 
 async function getPage(client: Client, pageId: string) {
-  const page = await client.pages.retrieve({ page_id: pageId });
-  return page as PageObjectResponse;
+    const page = await client.pages.retrieve({ page_id: pageId });
+    return page as PageObjectResponse;
 }
 
-async function main() {
-  const client = new Client({
-    auth: process.env.NOTION_TOKEN,
-  });
-
-  const gptscript = await import("@gptscript-ai/gptscript");
-  const gptscriptClient = new gptscript.GPTScript();
-
-  let output: OutputMetadata = {} as OutputMetadata;
-  let metadataFile;
-  try {
-    metadataFile = await gptscriptClient.readFileInWorkspace(".metadata.json");
-  } catch (err: any) {
-    // Ignore any error if the metadata file doesn't exist. Ideally we should check for only not existing error but sdk doesn't provide that
-  }
-  if (metadataFile) {
-    output = JSON.parse(metadataFile.toString());
-  }
-
-  if (!output.files) {
-    output.files = {};
-  }
-
-  if (!output.state) {
-    output.state = {} as {
-      notionState: {
-        pages: Record<
-          string,
-          { id: string; title: string; folderPath: string }
-        >;
-      };
-    };
-  }
+async function getAllPagesIteratively(
+    client: Client,
+    output: OutputMetadata,
+    gptscriptClient: any
+) {
+    const pages = new Map<
+        string,
+        {
+            page: PageObjectResponse;
+            path: string;
+        }
+    >();
+    const stack: (PageObjectResponse | null)[] = [];
 
-  if (!output.state.notionState) {
-    output.state.notionState = {} as {
-      pages: Record<string, { id: string; title: string; folderPath: string }>;
-    };
-  }
-
-  if (!output.state.notionState.pages) {
-    output.state.notionState.pages = {};
-  }
-
-  let syncedCount = 0;
-  const allPages = await client.search({
-    filter: { property: "object", value: "page" },
-  });
-
-  const pageUrls = new Set();
-  for (const page of allPages.results) {
-    let p = page as PageObjectResponse;
-    if (p.archived) {
-      continue;
+    let cursor = null;
+    do {
+        const response: SearchResponse = await client.search({
+            page_size: 100,
+            start_cursor: cursor ?? undefined,
+            filter: { property: "object", value: "page" },
+        });
+
+        stack.push(...(response.results as PageObjectResponse[]));
+
+        cursor = response.has_more ? response.next_cursor : null;
+    } while (cursor);
+
+    while (stack.length > 0) {
+        const currentPage = stack.pop();
+        if (!currentPage) {
+            continue;
+        }
+
+        if (pages.has(currentPage.id)) {
+            continue;
+        }
+        if (currentPage.url) {
+            output.status = `Syncing page ${currentPage.url}...`;
+            await gptscriptClient.writeFileInWorkspace(
+                ".metadata.json",
+                Buffer.from(JSON.stringify(output, null, 2))
+            );
```

Comment:
I'm confused, why is this file getting written over and over again, once for each page that has a URL?

Reply:
this is for writing the status back to the metadata file to provide some progress. The way it is done is that sync tool write status into a file in workspace and otto controller checks that in an interval and show status in api.